src/lib.rs

```diff
@@ -0,0 +1,171 @@
+extern crate bit_vec;
+extern crate num;
+extern crate rand;
+extern crate rustty;
+
+use std::collections::{HashSet};
+
+use bit_vec::BitVec;
+use num::bigint::BigInt;
+use rand::Rng;
+use rustty::ui::{Widget};
+use rustty::{CellAccessor};
+
+type Cell = (usize, usize);
+type CellSet = HashSet<Cell>;
+
+pub struct World {
+    height: usize,
+    width: usize,
+    rule: BitVec,
+    grid: CellSet,
+}
+
+impl World {
+    pub fn new((width, height): Cell, rule: BigInt) -> World {
+        let (_, bytes) = rule.to_bytes_be();
+        World {
+            height: height,
+            width: width,
+            rule: BitVec::from_bytes(&bytes),
+            grid: HashSet::with_capacity(height * width),
+        }
+    }
+
+    pub fn gen(&mut self) {
+        self.grid.clear();
+        for x in 1..self.width {
+            for y in 1..self.height {
+                if rand::thread_rng().gen_weighted_bool(30) {
+                    self.grid.insert((x, y));
+                }
+            }
+        }
+    }
+
+    fn get_present_state(&self, cell: &Cell) -> usize {
+
+        let (x, y) = (cell.0, cell.1);
+        let top = y.checked_sub(1) != None;
+        let bot = y.checked_add(1) <= Some(self.height);
+        let right = x.checked_add(1) <= Some(self.width);
+        let left = x.checked_sub(1) != None;
+
+        let mut val = 0;
+
+        if left && top {
+            val += (self.grid.contains(&(x-1, y-1)) as usize) << 0;
+        }
+        if top {
+            val += (self.grid.contains(&(x, y-1)) as usize)  << 1;
+        }
+        if top && right {
+            val += (self.grid.contains(&(x+1, y-1)) as usize) << 2;
+        }
+        if left {
+            val += (self.grid.contains(&(x-1, y)) as usize) << 3;
+        }
+        val += (self.grid.contains(&(x, y)) as usize) << 4;
+        if right {
+            val += (self.grid.contains(&(x+1, y)) as usize) << 5;
+        }
+        if left && bot {
+            val += (self.grid.contains(&(x-1, y+1)) as usize) << 6;
+        }
+        if bot {
+            val += (self.grid.contains(&(x, y+1)) as usize) << 7;
+        }
+        if bot && right {
+            val += (self.grid.contains(&(x+1, y+1)) as usize) << 8;
+        }
+        val
+    }
+
+    fn make_dead_or_alive(&self, cell: &Cell) -> bool {
+        let state = self.get_present_state(cell);
+        if state > self.rule.len() -1 {
+            return false
+        }
+        return self.rule[state];
+    }
+
+    // This is an obviously dumb way to do this
+    // TODO: Find a better way
+    fn neighbors(&self, cell: &Cell) -> CellSet {
```

Comment:
This function and `get_present_state()` have really similar logic, we should maybe work on joining them or at least adding a third function that they both call for neighbor validation.

Reply:
I agree. Making this an issue.